services/certificate/certificate-expiration.service.js

```diff
@@ -0,0 +1,110 @@
+import emojic from 'emojic'
+import Joi from 'joi'
+import { BaseService, pathParam, queryParam } from '../index.js'
+import trace from '../../core/base-service/trace.js'
+
+const queryParamSchema = Joi.object({
+  warningDays: Joi.number().min(0),
+  dangerDays: Joi.number().min(0),
+}).required()
+
+export default class CertificateExpiration extends BaseService {
+  static category = 'monitoring'
+
+  static route = {
+    base: 'certificate/expiration',
+    pattern: ':hostname',
+    queryParamSchema,
+  }
+
+  static openApi = {
+    '/certificate/expiration/{hostname}': {
+      get: {
+        summary: 'Certificate Expiration',
+        parameters: [
+          pathParam({ name: 'hostname', example: 'shields.io' }),
+          queryParam({
+            name: 'warningDays',
+            schema: { type: 'number' },
+            description:
+              'Warning (yellow) if less than this many days remain (default=30)',
+            example: '30',
+          }),
+          queryParam({
+            name: 'dangerDays',
+            schema: { type: 'number' },
+            description:
+              'Danger (red) if less than this many days remain (default=7)',
+            example: '7',
+          }),
+        ],
+      },
+    },
+  }
+
+  static _cacheLength = 0
+
+  async fetch({ hostname }) {
+    trace.logTrace('outbound', emojic.womanCook, 'trying', hostname)
+    let resp
+    try {
+      resp = await this._request({
+        url: `https://${hostname}/`,
+        options: {
+          followRedirect: false,
+        },
+      })
+      return { expiresStr: resp.res.req.socket.getPeerCertificate().valid_to }
+    } catch (err) {
+      // since got will throw on a redirect, we need to catch it here and possibly return the expiration date we found
+      const expiresStr =
+        err.response?.req?.socket?.getPeerCertificate()?.valid_to
+      trace.logTrace(
+        'outbound',
+        emojic.noGoodWoman,
+        'cert fetch error',
+        hostname,
+        expiresStr,
+        err,
+      )
+      if (!expiresStr) {
+        throw err
+      }
+      return { expiresStr }
+    }
+  }
+
+  async handle({ hostname }, { warningDays = 30, dangerDays = 7 }) {
+    try {
+      const { expiresStr } = await this.fetch({ hostname })
+      const expires = new Date()
+      expires.setTime(Date.parse(expiresStr))
+      const warningThreshold =
+        new Date().getTime() + warningDays * 24 * 60 * 60 * 1000
+      const dangerThreshold =
+        new Date().getTime() + dangerDays * 24 * 60 * 60 * 1000
+      const message = expires.toISOString().slice(0, 10)
```

Comment:
Can we use `formatRelativeDate()` https://github.com/badges/shields/blob/master/services/text-formatters.js#L185-L198 to display the date here.

Reply:
(and `isRelativeFormattedDate` for the tests assertions https://github.com/badges/shields/blob/master/services/test-validators.js#L119-L123 )